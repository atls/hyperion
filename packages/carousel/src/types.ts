export interface CarouselProps {
  transition: boolean
}
