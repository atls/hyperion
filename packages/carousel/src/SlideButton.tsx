import { layout, position } from 'styled-system'

import styled               from '@emotion/styled'

export const SlideButton = styled.div<any>(
  ({ theme, isSquareControls, disabled }: any) => ({
    position: 'absolute',
    background: theme.colors.white,
    boxSizing: 'border-box',
    borderRadius: isSquareControls ? '3px' : '50%',
    boxShadow: theme.shadows.controls,
    border: theme.borders.controls,
    zIndex: 9,
    display: disabled ? 'none' : 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    cursor: 'pointer',
    transition: '0.35s',
    '&:hover': {
      boxShadow: theme.shadows.woodsmoke,
    },
    '&:active': {
      transform: 'scale(0.9)',
    },
  }),
  position,
  layout,
)

SlideButton.defaultProps = {
  width: '64px',
  height: '64px',
  top: 'calc(50% - 32px)',
}
