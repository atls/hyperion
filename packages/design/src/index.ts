export * from './Sample'
export * from './Layout'
export * from './Container'
export * from './List'
export * from './Text'
export * from './Space'
