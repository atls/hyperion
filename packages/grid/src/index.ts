export * from './Column'
export * from './Grid'
