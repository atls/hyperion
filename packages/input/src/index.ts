export * from './Textarea'
export * from './Input'
export * from './Checkbox'
