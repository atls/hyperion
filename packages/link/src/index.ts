export * from './Link'
export * from './NextLink'
export * from './RouteLink'
