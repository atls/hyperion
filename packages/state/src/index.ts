export * from './withState'
export * from './useActive'
export * from './useHover'
export * from './useFocus'
