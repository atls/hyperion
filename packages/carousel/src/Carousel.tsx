import React, { useEffect, useRef, useState } from 'react'
import { Swipeable }                          from 'react-swipeable'
import { layout }                             from 'styled-system'

import styled                                 from '@emotion/styled'
import {
  ArrowBackwardIcon,
  ArrowForwardIcon,
  ArrowLeftIcon,
  ArrowRightIcon,
} from '@atlantis-lab/icons'
import { contentWidth, widthWithMargin }      from '@atlantis-lab/utils'

import { SlideButton }                        from './SlideButton'
import { CarouselProps }                      from './types'

const StyledCarousel = styled.div<CarouselProps>(({ transition }) => ({
  position: 'relative',
  display: 'flex',
  flexDirection: 'row',
  transition: transition ? '0.3s' : '',
  height: 'auto',
  '&:focus': {
    pointerEvents: 'none',
  },
  '&:hover': {
    cursor: 'pointer',
  },
}))

const Container = styled.div<any>(({ show }) => ({
  width: '100%',
  display: show ? 'flex' : 'none',
  position: 'relative',
}))

const Screen = styled.div(
  ({ isOverflowHidden }) => ({
    display: 'flex',
    overflow: isOverflowHidden ? 'hidden' : 'initial',
  }),
  layout
)

export const Carousel = ({
  children,
  disableButton = false,
  step = 1,
  controlWidth,
  controlHeight,
  controlRight,
  controlLeft,
  controlTop,
  isOverflowHidden,
  isSquareControls,
  theme,
}) => {
  const [enableTransition, setEnableTransition] = useState(true)
  const [innerWidth, setInnerWidth] = useState(null)
  const [fullWidth, setFullWidth] = useState(null)
  const [childWidth, setChildWidth] = useState(null)
  const [left, setLeft] = useState(0)
  const [deltaX, setDeltaX] = useState(0)
  const containerNode = useRef()
  const screenNode = useRef(null)

  const setWidth = () => {
    setInnerWidth(contentWidth(containerNode.current))
    setChildWidth(
      Array.prototype.map.call(screenNode.current && screenNode.current.children, item =>
        widthWithMargin(item)
      )
    )
  }

  useEffect(() => {
    setFullWidth(contentWidth(screenNode.current))
  })

  useEffect(() => {
    const timeout = setTimeout(() => {
      setWidth()
    }, 100)
    return () => clearTimeout(timeout)
  }, [fullWidth])

  useEffect(() => {
    if (!disableButton) {
      window.addEventListener('resize', setWidth)
      return () => window.removeEventListener('resize', setWidth)
    }
    return undefined
  }, [])

  const swiping = data => {
    if (innerWidth >= fullWidth) {
      return
    }
    const newLeft = deltaX - data.deltaX + left

    if (newLeft > 500) {
      setLeft(left)
      return
    }

    if (newLeft < innerWidth - fullWidth - 500) {
      setLeft(left)
      return
    }

    setDeltaX(data.deltaX)
    setLeft(newLeft)
    setEnableTransition(false)
  }

  const handleClick = direction => {
    let activeSlide = 0
    let activeLeft = 0
    for (;;) {
      if (left >= activeLeft && direction === 'left') break
      if (left > activeLeft && direction === 'right') break
      if (activeSlide > 100) break
      activeLeft -= childWidth[activeSlide]
      activeSlide += 1
    }
    setEnableTransition(true)
    let newLeft = 0
    if (direction === 'left') {
      childWidth.forEach((item, index) => {
        if (activeSlide - 1 - (step - 1) > index) {
          newLeft -= item
        }
      })
    }
    if (direction === 'right') {
      childWidth.forEach((item, index) => {
        if (activeSlide + (step - 1) > index) {
          newLeft -= item
        }
      })
    }

    if (left === 0 && newLeft >= 0) {
      setLeft(innerWidth - fullWidth)
      return
    }
    if (innerWidth - newLeft >= fullWidth) {
      if (left === innerWidth - fullWidth && innerWidth - newLeft > innerWidth - fullWidth) {
        setLeft(0)
        return
      }
      setLeft(innerWidth - fullWidth)
      return
    }

    setLeft(newLeft)
  }

  return (
    <Container show={children.length !== 0}>
      <SlideButton
        onClick={() => handleClick('left')}
        disabled={disableButton}
        width={controlWidth}
        height={controlHeight}
        left={controlLeft}
        top={controlTop}
        isSquareControls={isSquareControls}
        theme={theme}
      >
        {!isSquareControls && <ArrowBackwardIcon width={24} height={24} />}
        {isSquareControls && <ArrowLeftIcon width={20} height={10} />}
      </SlideButton>
      <Screen ref={containerNode} isOverflowHidden={isOverflowHidden} maxWidth='100%'>
        <Swipeable
          onSwiping={data => swiping(data)}
          onSwiped={() => {
            setDeltaX(0)
            if (left > 0) {
              setEnableTransition(true)
              setLeft(0)
              if (left === 0) setEnableTransition(false)
            }

            if (left <= innerWidth - fullWidth) {
              setEnableTransition(true)
              setLeft(innerWidth - fullWidth)
              if (left === innerWidth - fullWidth) setEnableTransition(false)
            }
          }}
          preventDefaultTouchmoveEvent
          trackMouse
          trackTouch
          delta={15}
        >
          <StyledCarousel ref={screenNode} style={{ left }} transition={enableTransition}>
            {children}
          </StyledCarousel>
        </Swipeable>
      </Screen>
      <SlideButton
        onClick={() => handleClick('right')}
        disabled={disableButton}
        width={controlWidth}
        height={controlHeight}
        right={controlRight}
        top={controlTop}
        isSquareControls={isSquareControls}
        theme={theme}
      >
        {!isSquareControls && <ArrowForwardIcon width={24} height={24} />}
        {isSquareControls && <ArrowRightIcon width={20} height={10} />}
      </SlideButton>
    </Container>
  )
}
