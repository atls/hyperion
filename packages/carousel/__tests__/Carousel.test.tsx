import React        from 'react'
import renderer     from 'react-test-renderer'

import { Carousel } from '../src/index'

it('Carousel should render correctly', () => {
  const tree = renderer
    .create(
      <Carousel
        controlWidth={50}
        controlHeight={50}
        controlRight={0}
        controlLeft={0}
        controlTop={100}
        isOverflowHidden={false}
        isSquareControls={false}
      >
        <h1>{'bigtext'.repeat(100)}</h1>
      </Carousel>,
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})
