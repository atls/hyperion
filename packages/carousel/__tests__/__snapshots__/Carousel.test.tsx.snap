// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel should render correctly 1`] = `
.emotion-4 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
}

.emotion-0 {
  position: absolute;
  background: #fff;
  box-sizing: border-box;
  border-radius: 50%;
  box-shadow: 0 0 10px rgba(0,0,0,0.5);
  border: 2px solid black;
  z-index: 9;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  -webkit-transition: 0.35s;
  transition: 0.35s;
  left: 0;
  top: 100px;
  width: 50px;
  height: 50px;
}

.emotion-0:active {
  -webkit-transform: scale(0.9);
  -ms-transform: scale(0.9);
  transform: scale(0.9);
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  overflow: initial;
  max-width: 100%;
}

.emotion-1 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-transition: 0.3s;
  transition: 0.3s;
  height: auto;
}

.emotion-1:focus {
  pointer-events: none;
}

.emotion-1:hover {
  cursor: pointer;
}

.emotion-3 {
  position: absolute;
  background: #fff;
  box-sizing: border-box;
  border-radius: 50%;
  box-shadow: 0 0 10px rgba(0,0,0,0.5);
  border: 2px solid black;
  z-index: 9;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  -webkit-transition: 0.35s;
  transition: 0.35s;
  right: 0;
  top: 100px;
  width: 50px;
  height: 50px;
}

.emotion-3:active {
  -webkit-transform: scale(0.9);
  -ms-transform: scale(0.9);
  transform: scale(0.9);
}

<div
  className="emotion-4"
>
  <div
    className="emotion-0"
    disabled={false}
    height={50}
    onClick={[Function]}
    width={50}
  >
    <svg
      fill="none"
      height={24}
      viewBox="0 0 24 24"
      width={24}
    >
      <mask
        height={16}
        id="ArrowBackwardIcon"
        mask-type="alpha"
        maskUnits="userSpaceOnUse"
        width={16}
        x={4}
        y={4}
      >
        <path
          clipRule="evenodd"
          d="M13.4142 5.41425L12 4L4 12L12 20L13.4142 18.5858L7.81634 13H20V11H7.81635L13.4142 5.41425Z"
          fill="white"
          fillRule="evenodd"
        />
      </mask>
      <g
        mask="url(#ArrowBackwardIcon)"
      >
        <rect
          fill="#999999"
          height={80}
          width={80}
          x={-28}
          y={-28}
        />
      </g>
    </svg>
  </div>
  <div
    className="emotion-2"
  >
    <div
      onMouseDown={[Function]}
    >
      <div
        className="emotion-1"
        style={
          Object {
            "left": 0,
          }
        }
      >
        <h1>
          bigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtextbigtext
        </h1>
      </div>
    </div>
  </div>
  <div
    className="emotion-3"
    disabled={false}
    height={50}
    onClick={[Function]}
    width={50}
  >
    <svg
      fill="none"
      height={24}
      viewBox="0 0 24 24"
      width={24}
    >
      <mask
        height={16}
        id="ArrowForwardIcon"
        mask-type="alpha"
        maskUnits="userSpaceOnUse"
        width={16}
        x={4}
        y={4}
      >
        <path
          clipRule="evenodd"
          d="M10.5858 5.41425L12 4L20 12L12 20L10.5858 18.5858L16.1837 13H4V11H16.1837L10.5858 5.41425Z"
          fill="white"
          fillRule="evenodd"
        />
      </mask>
      <g
        mask="url(#ArrowForwardIcon)"
      >
        <rect
          fill="#999999"
          height={80}
          width={80}
          x={-28}
          y={-28}
        />
      </g>
    </svg>
  </div>
</div>
`;
